relicensing blog post (#250)
MDAnalysis/MDAnalysis.github.io: _posts/2022-11-02-relicensing.md

---
layout: post
title: Relicensing MDAnalysis
---

<img src="https://www.gnu.org/graphics/lgplv3-with-text-154x68.png"
title="LGPLv3" alt="LGPLv3 logo" style="float: right"/>

We want to **change the license of MDAnalysis** from the [GNU General

> Public License][GPLv2] (GPL v2+) to the more permissive [Lesser GNU

Reviewer: todo - don't call this permissive

Author: Why not?

Reviewer: permissive is a very specific category under licensing, imho just bringing it up opens you up to further scrutiny. It may be that this is the only word we can use here, but I'd like to give it some thought ahead of using such specific language.

Author: ok, if this is too loaded then we don't want to stir up controversy, and of course not overclaim.

Lesser Public License][LGPL] (LGPL v3+) license. Both are *open source
licenses* but the [LGPL][] gives developers more freedom in how they
license any of their own code that makes use of MDAnalysis.

The license is the legal basis that determines how code can be copied,
changed, and incorporated into other code. The license is the only
reason that one can use someone else's code --- it does not matter if
the code is publicly available somewhere. If it does not have a
license you cannot use it legally. Therefore, getting the license

> right is very important.

Reviewer: to-do: reword tone

**In order to change the MDAnalysis license, we need the agreement of

> all our contributors to change from GPL to LGPL.**

Reviewer: todo: also detail how this is a community consultation, if the MDA users don't want this then we won't go ahead

In this post we want to share the motivation, outline the relicensing

> process, and invite questions from the community.

Reviewer: todo: point to the new licensing email address

## Why not GPL any more?

Since its initial release in 2008, MDAnalysis has grown from a small
Python package used by a handful of enthusiastic graduate students and
postdocs to a mature library that is used by thousands of researchers
in the molecular sciences. The MDAnalysis library was published under
an open source license from the start so that anyone could freely use
it, contribute to it, and build on it. We chose the GNU Public License
(GPL) for this purpose. Thanks to the "copy-left" portion of the GPL,
which requires anyone using MDAnalysis in their own code to also adopt
the GPL for their code, contributors could feel that any time and work
that they donated and invested into MDAnalysis would not end up
contributing to software that they could not use themselves in the

> future.

Reviewer: todo: remove or reword, this is slightly misleading since GPL code can be used in commercial settings

Author: Maybe I am misunderstanding your concern. This does not explicitly talk about commercial code. It simply iterates the guarantee of the GPL that code remains free. If the commercial code is under GPL I get to use it under GPL.

Reviewer: > would not end up
contributing to software that they could not use themselves

This is the problematic part of the sentence. There is absolutely nothing under GPL that prevents folks from taking your code and using it for their internal purposes. Also code being under GPL doesn't mean it can't be impossible to access, GPL doesn't mean you have a legal obligation to release (so technically it could it behind a SaaS wall or anything else)

Author: I understand that there's no requirement to release the code and SaaS is a general problem. But for the cases where you get your hands on code (which is easier for interpreted languages than compiled ones), GPL seems the clearest way to ensure that the code remains accessible. 

I think one problem we have with this post is if it needs to be written for lawyers or for users/developers. Ideally for both... but if I have to come down on one side, then users/developers. You rightly point out that we need to be careful with our language.

Reviewer: hopefully if we make it real clear in a disclaimer at the start then it should be fine, but it's one of those things that someone might look at one day and go "but MDAnalysis said this"

I'm currently considering some very bolded wording that says "we aren't lawyers, this is not legal advice, if you want it please contact a lawyer"

Author: Loved your disclaimer in PR #251

However, the GPL has also created barriers to adoption of MDAnalysis

> in that it prevented developers to make their own code available under

Reviewer 2: ```suggestion
in that it prevented developers from making their own code available under
```

other permissive open source licenses. Ultimately, we do not want to

> tell our developers and users how to license their code but we also

Reviewer: todo: reword as our own legal interpretation, we don't want to be getting into a fight with folks out there that believe a different interpretation

Author: For everything that we are saying we should have a disclaimer that this our interpretation. However, I have no problem defending our interpretation.

Reviewer: > I have no problem defending our interpretation.

Generally INAL should stand here, we have taken advice from a lawyer and we are sticking with it. I would always advise making sure that language is as soft as possible in order to not have to suddenly take a stance as a non-legal professional.

want to clearly signal that the work on MDAnalysis itself is

> guaranteed to remain open and free.

Reviewer 2: Should we explicitly state some more of the benefits of re-licensing, e.g pave way for MDAKits,  grow our user and contributor base open up avenues for greater adoption in industry etc. ?

Author: I think it would be good to be open about our motivations. Draft something concise!

Reviewer: Should be addressed, please have a look (I kept it small).

## Why now the LGPL?

We therefore decided to start the process to **relicense MDAnalysis
under the Lesser GNU Public License v3 (LGPL)**. This open source

> license fulfills a number of important requirements for us:

Author: I think one disclaimer is sufficient.

1. Downstream codes are able to freely import or link to MDAnalysis
   library components without impacting the license choice of the
   downstream code.
   
2. Downstream codes are able to use and subclass any MDAnalysis components
   under its application programming interface (classes, methods, and
   data objects), without impacting the license choice of the
   downstream code.
   
3. Codes which either copy or extend the MDAnalysis library should
   fall under the copyleft license requirements of the MDAnalysis
   library license.
   
Thus, the LGPL gives people the freedom to choose any license for
their own code that *makes use* of the MDAnalysis library as a whole
(namely ``import MDAnalysis`` or subclassing). They can even make a
commercial product out of it with a restrictive license (although we
encourage open source licenses). However, one cannot just take parts
of the MDAnalysis code and put it into another piece of code unless
the other code is then *also* licensed under the GPL/LGPL. The LGPL
ensures that the MDAnalysis library itself is guaranteed to remain

> open and free.

Reviewer: I don't think we can necessarily make that promise, todo: reword

Author: But it is our intention, is it not? I'd say if you weaken this statement then you invite questions as to what sinister plans we might be hatching.

Reviewer: This is the realm of legal problems @orbeckst - we should not _guarantee_ anything (it's annoying I know). We can take a strong stance that this is what we will do, but don't open yourself to legal trouble.

Author: Fair.

> We considered other licenses but none fulfilled our requirements.

Reviewer: todo: reword, we don't want to say that the conversation is dead, otherwise why have a community discussion

Author: Well.... we are not looking for a discussion about what alternative license we should take. In my view, we did our due diligence and came up with a solution. This is more about building understanding and explaining.

It is more — as you wanted to add above — if the community does not like LGPL then we will remain GPL.

Author: If there's any disagreement on this point amonst @MDAnalysis/coredevs then we better get on a call soon.

Reviewer: @orbeckst I'm not sure it's a disagreement, with all the other points, it's a case of adjusting language to not shut the door. I'm not saying this specific sentence needs rewording (this is a todo, otherwise I'd just have rewritten it there and then), but it's a "something that needs to be considered when considering the tone of this section" (this is true for all other todos here)

   
## How will the relicensing process work?

MDAnalysis has over 150 contributors. They all contributed code under
the GPL v2+. In order to relicense, we need *all our
contributors to agree to the license change*. 

> We know that this is a big change for the MDAnalysis community (and

Author: absolutely!!

it's a big administrative and legal undertaking for us). We therefore
want to give everyone an opportunity to ask questions so that you can
make an informed decision.

### Consultation period

For the next month, the MDAnalysis Core Developers will answer your
questions regarding the relicensing in our [public forums]({% link

> index.md %}#participating) (mailing lists, discord, twitter).

Author: Yes, very good, I had forgotten about it.

### Contacting contributors

After the consultation period, we will contact every contributor to
MDAnalysis with a request to agree to changing the license from the
current "GPL v2 or any later version" to "LGPL v3 or any later
version".

### License change

Once we have the necessary agreement, we will switch over the license

> of MDAnalysis.

Author: Fair enough. I just thought we need to somwhere say that we actually do something with the license.

### Timeline

* November 2022: consultation period
* December 2022: start contacting contributors
* 2023?: We do not know how long the process will take. Our aim is to
change the license as quickly as possible. We will keep the community

> updated.

Reviewer: Todo - make sure we give a final date for the consultation period before we submit this

Author: yes
